File: apps/posting/posting_core/clients/x.py
from __future__ import annotations

import os

from posting_core.publish_config import X_ACCESS_TOKEN, X_ACCESS_TOKEN_SECRET, X_CONSUMER_KEY, X_CONSUMER_SECRET, log
from posting_core.text import strip_urls

def upload_x_media_chunked(file_path, auth, media_type="video/mp4", media_category="tweet_video"):
    import os
    import time
    import requests
    
    file_size = os.path.getsize(file_path)
    upload_url = "https://upload.twitter.com/1.1/media/upload.json"
    
    # 1. INIT
    log(f"X Media Chunked Upload [INIT] for size {file_size} bytes...")
    init_data = {
        "command": "INIT",
        "total_bytes": file_size,
        "media_type": media_type,
        "media_category": media_category
    }
    res = requests.post(upload_url, auth=auth, data=init_data, timeout=30)
    if not (200 <= res.status_code < 300):
        raise Exception(f"X INIT failed: {res.status_code} {res.text}")
    
    media_data = res.json()
    media_id_str = media_data["media_id_string"]
    log(f"X Media ID allocated: {media_id_str}")
    
    # 2. APPEND
    chunk_size = 2 * 1024 * 1024  # 2MB chunks
    segment_index = 0
    
    with open(file_path, "rb") as f:
        while True:
            chunk_data = f.read(chunk_size)
            if not chunk_data:
                break
                
            log(f"X Media Chunked Upload [APPEND] chunk {segment_index} ({len(chunk_data)} bytes)...")
            append_data = {
                "command": "APPEND",
                "media_id": media_id_str,
                "segment_index": segment_index
            }
            files = {"media": chunk_data}
            res = requests.post(upload_url, auth=auth, data=append_data, files=files, timeout=60)
            if not (200 <= res.status_code < 300):
                raise Exception(f"X APPEND chunk {segment_index} failed: {res.status_code} {res.text}")
                
            segment_index += 1
            
    # 3. FINALIZE
    log("X Media Chunked Upload [FINALIZE]...")
    finalize_data = {
        "command": "FINALIZE",
        "media_id": media_id_str
    }
    res = requests.post(upload_url, auth=auth, data=finalize_data, timeout=30)
    if not (200 <= res.status_code < 300):
        raise Exception(f"X FINALIZE failed: {res.status_code} {res.text}")
        
    finalize_res = res.json()
    
    # 4. STATUS check (only if processing_info is in response)
    processing_info = finalize_res.get("processing_info")
    if processing_info:
        state = processing_info.get("state")
        log(f"X Media processing state: {state}")
        
        while state in ("pending", "in_progress"):
            check_after_secs = processing_info.get("check_after_secs", 5)
            log(f"Waiting {check_after_secs} seconds for X video processing...")
            time.sleep(check_after_secs)
            
            status_params = {
                "command": "STATUS",
                "media_id": media_id_str
            }
            res = requests.get(upload_url, auth=auth, params=status_params, timeout=30)
            if not (200 <= res.status_code < 300):
                raise Exception(f"X STATUS check failed: {res.status_code} {res.text}")
                
            status_res = res.json()
            processing_info = status_res.get("processing_info") or {}
            state = processing_info.get("state")
            log(f"X Media processing state update: {state}")
            
            if state == "failed":
                error_info = processing_info.get("error", {})
                raise Exception(f"X video processing failed: {error_info.get('message', 'Unknown error')}")
                
    log("X Media upload and processing completed successfully.")
    return media_id_str


def publish_to_x(text, media_items):
    if not X_CONSUMER_KEY or not X_CONSUMER_SECRET or not X_ACCESS_TOKEN or not X_ACCESS_TOKEN_SECRET:
        log("X/Twitter credentials missing")
        return None
        
    try:
        import requests
        from requests_oauthlib import OAuth1
    except ImportError:
        log("Error: requests or requests-oauthlib not installed")
        return None
        
    auth = OAuth1(X_CONSUMER_KEY, X_CONSUMER_SECRET, X_ACCESS_TOKEN, X_ACCESS_TOKEN_SECRET)
    text = strip_urls(text)
    
    media_ids = []
    try:
        # 1. Загрузка медиа (если есть)
        for item in media_items:
            file_path = item.get("local_path")
            if not file_path or not os.path.exists(file_path):
                log(f"Local file for X upload not found: {file_path}")
                continue
                
            item_type = item.get("type")
            try:
                if item_type == "VIDEO":
                    media_id_str = upload_x_media_chunked(file_path, auth, media_type="video/mp4", media_category="amplify_video")
                else:
                    log(f"Uploading image to X: {file_path}")
                    upload_url = "https://upload.twitter.com/1.1/media/upload.json"
                    with open(file_path, "rb") as f:
                        files = {"media": f}
                        res = requests.post(upload_url, auth=auth, files=files, timeout=60)
                        
                    if res.status_code not in (200, 201):
                        log(f"X media upload failed: {res.status_code} {res.text}")
                        continue
                    media_data = res.json()
                    media_id_str = media_data.get("media_id_string")
                
                if media_id_str:
                    media_ids.append(media_id_str)
                    log(f"Uploaded X media ID: {media_id_str}")
            except Exception as exc:
                log(f"Failed to upload media to X ({item_type}): {exc}")
                
        # 2. Создание твита
        tweet_url = "https://api.twitter.com/2/tweets"
        payload = {"text": text}
        if media_ids:
            payload["media"] = {"media_ids": media_ids}
            
        log("Creating tweet on X...")
        res = requests.post(tweet_url, auth=auth, json=payload, timeout=30)
        if res.status_code in (200, 201):
            tweet_data = res.json()
            tweet_id = tweet_data.get("data", {}).get("id")
            log(f"X tweet published successfully: {tweet_id}")
            return tweet_id
        else:
            log(f"Failed to create tweet: {res.status_code} {res.text}")
            return None
            
    except Exception as exc:
        log(f"Error publishing to X: {exc}")
        return None

File: apps/backend/src/social/x.ts
import crypto from "node:crypto";
import fs from "node:fs";
import OAuth from "oauth-1.0a";
import type { BackendConfig } from "../config.js";
import type { PublishResult } from "../publishing/errors.js";
import { HttpPublishError } from "../publishing/errors.js";
import { guessContentType, payloadMedia, payloadText, stripUrls } from "./payload.js";

const UPLOAD_URL = "https://upload.twitter.com/1.1/media/upload.json";

export async function publishToX(
  payload: Record<string, unknown>,
  config: BackendConfig,
  fetchImpl: typeof fetch = fetch,
): Promise<PublishResult> {
  assertCredentials(config);
  const mediaIds: string[] = [];
  for (const item of payloadMedia(payload)) {
    if (!item.localPath || !fs.existsSync(item.localPath)) continue;
    mediaIds.push(
      item.type === "VIDEO" ? await uploadVideo(item.localPath, config, fetchImpl) : await uploadImage(item.localPath, config, fetchImpl),
    );
  }
  const body = JSON.stringify({ text: stripUrls(payloadText(payload)), ...(mediaIds.length ? { media: { media_ids: mediaIds } } : {}) });
  const response = await oauthFetch("https://api.twitter.com/2/tweets", config, fetchImpl, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body,
  });
  const result = await jsonResponse<{ data?: { id?: string } }>(response, "X tweet create");
  const id = result.data?.id;
  return { ok: Boolean(id), id: id ?? null, url: id ? `https://x.com/i/web/status/${id}` : null, raw: result };
}

async function uploadImage(filePath: string, config: BackendConfig, fetchImpl: typeof fetch): Promise<string> {
  const form = new FormData();
  form.set(
    "media",
    new Blob([await fs.promises.readFile(filePath)], { type: guessContentType(filePath) }),
    filePath.split("/").pop() || "image",
  );
  const response = await oauthFetch(UPLOAD_URL, config, fetchImpl, { method: "POST", body: form });
  const result = await jsonResponse<{ media_id_string?: string }>(response, "X media upload");
  if (!result.media_id_string) throw new Error("X media upload missing media_id_string");
  return result.media_id_string;
}

async function uploadVideo(filePath: string, config: BackendConfig, fetchImpl: typeof fetch): Promise<string> {
  const initParams = new URLSearchParams({
    command: "INIT",
    total_bytes: String(fs.statSync(filePath).size),
    media_type: "video/mp4",
    media_category: "amplify_video",
  });
  const initialized = await jsonResponse<{ media_id_string?: string }>(
    await oauthFetch(UPLOAD_URL, config, fetchImpl, formInit(initParams), initParams),
    "X media INIT",
  );
  const mediaId = initialized.media_id_string;
  if (!mediaId) throw new Error("X media INIT missing media_id_string");

  const handle = await fs.promises.open(filePath, "r");
  try {
    let position = 0;
    let segmentIndex = 0;
    const chunk = Buffer.alloc(2 * 1024 * 1024);
    while (true) {
      const { bytesRead } = await handle.read(chunk, 0, chunk.length, position);
      if (bytesRead === 0) break;
      const form = new FormData();
      form.set("command", "APPEND");
      form.set("media_id", mediaId);
      form.set("segment_index", String(segmentIndex));
      form.set("media", new Blob([chunk.subarray(0, bytesRead)], { type: "application/octet-stream" }), `segment-${segmentIndex}`);
      const response = await oauthFetch(UPLOAD_URL, config, fetchImpl, { method: "POST", body: form });
      if (!response.ok) throw await responseError(response, `X media APPEND ${segmentIndex}`);
      position += bytesRead;
      segmentIndex += 1;
    }
  } finally {
    await handle.close();
  }

  const finalizeParams = new URLSearchParams({ command: "FINALIZE", media_id: mediaId });
  const finalized = await jsonResponse<ProcessingResponse>(
    await oauthFetch(UPLOAD_URL, config, fetchImpl, formInit(finalizeParams), finalizeParams),
    "X media FINALIZE",
  );
  await waitForProcessing(mediaId, finalized.processing_info, config, fetchImpl);
  return mediaId;
}

async function waitForProcessing(
  mediaId: string,
  initial: ProcessingInfo | undefined,
  config: BackendConfig,
  fetchImpl: typeof fetch,
): Promise<void> {
  let processing = initial;
  const deadline = Date.now() + 600_000;
  while (processing && ["pending", "in_progress"].includes(processing.state ?? "")) {
    if (Date.now() >= deadline) throw new Error("X media processing timeout");
    await delay(Math.max(1, processing.check_after_secs ?? 5) * 1000);
    const query = new URLSearchParams({ command: "STATUS", media_id: mediaId });
    const result = await jsonResponse<ProcessingResponse>(
      await oauthFetch(`${UPLOAD_URL}?${query}`, config, fetchImpl, { method: "GET" }),
      "X media STATUS",
    );
    processing = result.processing_info;
    if (processing?.state === "failed") throw new Error(`X video processing failed: ${processing.error?.message ?? "Unknown error"}`);
  }
}

type ProcessingInfo = { state?: string; check_after_secs?: number; error?: { message?: string } };
type ProcessingResponse = { processing_info?: ProcessingInfo };

function formInit(params: URLSearchParams): RequestInit {
  return { method: "POST", headers: { "Content-Type": "application/x-www-form-urlencoded" }, body: params };
}

async function oauthFetch(
  url: string,
  config: BackendConfig,
  fetchImpl: typeof fetch,
  init: RequestInit,
  formParams?: URLSearchParams,
): Promise<Response> {
  const method = (init.method ?? "GET").toUpperCase();
  const authorization = oauthAuthorization(method, url, config, formParams);
  return fetchImpl(url, { ...init, headers: { ...init.headers, Authorization: authorization } });
}

export function oauthAuthorization(
  method: string,
  rawUrl: string,
  config: BackendConfig,
  formParams?: URLSearchParams,
  nonce = crypto.randomBytes(16).toString("hex"),
  timestamp = Math.floor(Date.now() / 1000),
): string {
  const credentials = xCredentials(config);
  const oauth = new OAuth({
    consumer: { key: credentials.consumerKey, secret: credentials.consumerSecret },
    signature_method: "HMAC-SHA1",
    hash_function: (base, key) => crypto.createHmac("sha1", key).update(base).digest("base64"),
  });
  oauth.getNonce = () => nonce;
  oauth.getTimeStamp = () => timestamp;
  const data = formParams ? Object.fromEntries(formParams.entries()) : undefined;
  const authorization = oauth.authorize(
    { url: rawUrl, method: method.toUpperCase(), ...(data ? { data } : {}) },
    {
      key: credentials.accessToken,
      secret: credentials.accessTokenSecret,
    },
  );
  return oauth.toHeader(authorization).Authorization;
}

async function jsonResponse<T>(response: Response, label: string): Promise<T> {
  const body = await response.text();
  if (!response.ok) throw new HttpPublishError(`${label} ${response.status}: ${body}`, response.status, body);
  return body ? (JSON.parse(body) as T) : ({} as T);
}

async function responseError(response: Response, label: string): Promise<HttpPublishError> {
  const body = await response.text();
  return new HttpPublishError(`${label} ${response.status}: ${body}`, response.status, body);
}

function assertCredentials(config: BackendConfig): void {
  void xCredentials(config);
}

function xCredentials(config: BackendConfig): {
  consumerKey: string;
  consumerSecret: string;
  accessToken: string;
  accessTokenSecret: string;
} {
  if (!config.X_CONSUMER_KEY || !config.X_CONSUMER_SECRET || !config.X_ACCESS_TOKEN || !config.X_ACCESS_TOKEN_SECRET)
    throw new Error("missing X credentials");
  return {
    consumerKey: config.X_CONSUMER_KEY,
    consumerSecret: config.X_CONSUMER_SECRET,
    accessToken: config.X_ACCESS_TOKEN,
    accessTokenSecret: config.X_ACCESS_TOKEN_SECRET,
  };
}

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
